x-pack/plugin/mapper-match-only-text/src/main/java/org/elasticsearch/xpack/matchonlytext/query/SourceConfirmedTextQuery.java

```diff
@@ -0,0 +1,345 @@
+                if (approximationScorer == null) {
+                    return null;
+                }
+                final DocIdSetIterator approximation = approximationScorer.iterator();
+                final LeafSimScorer leafSimScorer = new LeafSimScorer(simScorer, context.reader(), field, scoreMode.needsScores());
+                final CheckedIntFunction<List<Object>, IOException> valueFetcher = valueFetcherProvider.apply(context);
+                return new RuntimePhraseScorer(this, approximation, leafSimScorer, valueFetcher, field, in);
+            }
+
+        };
+    }
+
+    private class RuntimePhraseScorer extends Scorer {
+
+        private final LeafSimScorer scorer;
+        private final CheckedIntFunction<List<Object>, IOException> valueFetcher;
+        private final String field;
+        private final Query query;
+        private final TwoPhaseIterator twoPhase;
+
+        private int doc = -1;
+        private float freq;
+
+        private RuntimePhraseScorer(
+            Weight weight,
+            DocIdSetIterator approximation,
+            LeafSimScorer scorer,
+            CheckedIntFunction<List<Object>, IOException> valueFetcher,
+            String field,
+            Query query
+        ) {
+            super(weight);
+            this.scorer = scorer;
+            this.valueFetcher = valueFetcher;
+            this.field = field;
+            this.query = query;
+            twoPhase = new TwoPhaseIterator(approximation) {
+
+                @Override
+                public boolean matches() throws IOException {
+                    return freq() > 0;
+                }
+
+                @Override
+                public float matchCost() {
+                    // TODO what is a right value?
+                    // Defaults to a high-ish value so that it likely runs last.
+                    return 10_000f;
+                }
+
+            };
+        }
+
+        @Override
+        public DocIdSetIterator iterator() {
+            return TwoPhaseIterator.asDocIdSetIterator(twoPhaseIterator());
+        }
+
+        @Override
+        public TwoPhaseIterator twoPhaseIterator() {
+            return twoPhase;
+        }
+
+        @Override
+        public float getMaxScore(int upTo) throws IOException {
+            return scorer.getSimScorer().score(Float.MAX_VALUE, 1L);
+        }
+
+        @Override
+        public float score() throws IOException {
+            return scorer.score(docID(), freq());
+        }
+
+        @Override
+        public int docID() {
+            return twoPhase.approximation().docID();
+        }
+
+        private float freq() throws IOException {
+            if (doc != docID()) {
+                doc = docID();
+                freq = computeFreq();
+            }
+            return freq;
+        }
+
+        private float computeFreq() throws IOException {
+            MemoryIndex index = new MemoryIndex();
```

Comment:
Sneaky, I like it.  MemoryIndex always runs with ScoreMode.COMPLETE so it might be possible to speed things up a bit by calling `createIndexSearcher` and running against that with ScoreMode.COMPLETE_NO_SCORES.  Another trick that highlighters use when re-analyzing source is to add a FilteringTokenFilter to the MemoryIndex's analyzer that drops all tokens not included in the query - you need to be careful to catch Automata-based queries but QueryVisitor makes it not too difficult to implement.

Reply:
This feels like a great idea for a follow-up PR? 0:)